### ISSUE TITLE
Enable Access to Dependency Request Objects in DependencyCollector

### ISSUE THREAD

**Issue by michaelwstark** [Microsoft/ApplicationInsights-dotnet-server#900]
Opening this item per comments on the pull request -- consolidating discussions referenced items.

#issues:
https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/820
https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/747
https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/587
https://github.com/Microsoft/ApplicationInsights-dotnet-server/pull/897

Please reference the issues and pull request for additional details and\or discussion.

Consumers of the dependency collection package would like to be able to populate additional data in the tracked telemetry through implementing ITelemetryInitializer classes.  At present, this is not possible given only the data passed in the call to Initialize() is the dependency telemetry item itself with pre-defined fields populated from within the dependency collector.

The ask is to track key data in the dependency collection and provide access to it as part of the ITelemetry item passed to the initializers.

Here is a sample of intended use -- reporting a debug-id value from an http response:

```csharp
public class SampleTelemetryInitializer : ITelemetryInitializer
{
    private const string DebugHeaderName = "debug-id";

    public void Initialize(ITelemetry telemetry)
    {
        var dependencyTelemetry = telemetry as DependencyTelemetry;
        if (dependencyTelemetry != null && dependencyTelemetry.TryGetOperationDetail("HttpResponse", out var reponseObject))
        {
            IEnumerable<string> debugData = null;

            var response = reponseObject as HttpResponseMessage;
            if (response?.Headers?.TryGetValues(DebugHeaderName, out debugData) ?? false)
            {
                if (debugData != null)
                {
                    dependencyTelemetry.Properties.Add(DebugHeaderName, string.Join(",", debugData));
                }
            }
        }
    }
}
```

**Comment by pschaeflein** [comment]
The issue mentioned by @christianacca is happening to me.

@lmolkova Should this issue be closed? Is there a different issue tracking a but rather than the implementation?

**Comment by pschaeflein** [comment]
Yes, Initialize is being called twice, however neither call includes ResponseHeaders. (The rawObjects field always contains only 1 item, the request object.)

I will try to get a repro together. I'm running a bot hosted locally at the moment. (Production will be on Azure App Service.) Below is a list of the NuGet dependencies. Should this environment be supported?
 
![image](https://user-images.githubusercontent.com/8008013/51181282-662b9400-1890-11e9-9b2d-39f0f22eed4f.png)

**Comment by michaelwstark** [comment]
@pschaeflein When you say that it is hosted locally -- what is the hosting process?  E.g. IIS, IIS Express or an Azure Emulator of some kind, etc

**Comment by pschaeflein** [comment]
Sorry, IISExpress/Win10

**Comment by cijothomas** [comment]
The response headers are read and populated correctly in .NET Framework.
https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/develop/Src/DependencyCollector/Shared/Implementation/HttpDesktopDiagnosticSourceListener.cs#L86
https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/develop/Src/DependencyCollector/Shared/Implementation/HttpProcessing.cs#L385

However, I dont see the same for .NET Core.
https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/develop/Src/DependencyCollector/Shared/HttpCoreDiagnosticSourceListener.cs

@michaelwstark Any chance you can take a look? Was it intentional or just happened to have left out for .net core?

**Comment by lmolkova** [comment]
@cijothomas 

ResponseHeaders are set when response was received, but response object is not available - this is .NET Fx case only.

On .NET Core response object is always available (if response was received). It is set here
-  for netcore2.0:  
https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/develop/Src/DependencyCollector/Shared/HttpCoreDiagnosticSourceListener.cs#L398
- for netcore 1.1 and 1.0
https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/develop/Src/DependencyCollector/Shared/HttpCoreDiagnosticSourceListener.cs#L463


@pschaeflein  could it be the case that there were no response (DNS issue, timeout exception, etc)?

**Comment by michaelwstark** [comment]
@pschaeflein  My personal experience during development and integration with my services was that I was never able to get anything set when running in a IIS Express host process.  Not sure what the key differences are\were (or whether the conclusion is entirely accurate), but the minute I actually deployed it to an development environment running in IIS everything worked as expected.  Not ideal.  I'll sanity check that this continues to be true for me later today and report what version my services are currently consuming.  I did just confirm that all the telemetry is reporting accurately in all of our non-IIS Express environments though (and has been very useful).

@lmolkova \ @cijothomas - Any idea why these values would be null in IIS Express from your domain knowledge?
